sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java

```diff
@@ -494,7 +495,22 @@ protected final void removeTemporaryFiles(List<String> knownFiles, PipelineOptio
       LOG.debug("Removing temporary bundle output files in {}.", tempDirectory);
       FileOperations fileOperations =
           FileOperationsFactory.getFileOperations(tempDirectory, options);
-      fileOperations.removeDirectoryAndFiles(tempDirectory, knownFiles);
+      IOChannelFactory factory = IOChannelUtils.getFactory(tempDirectory);
+
+      // To partially mitigate the effects of filesystems with eventually-consistent
+      // directory matching APIs, takes a list of files that are known to exist - i.e. removes the
```

Comment:
Nitpick: the comment was written for a function and thus phrased as "does X". But inside code, this is not the common style. The comment should not be transferred verbatim.

Rephrase slightly:
To partially mitigate the effects of filesystems with eventually-consistent directory matching APIs, we remove not only files that the filesystem says exist in the directory (which may be incomplete), but also files that are known to exist (produced by successfully completed bundles). This may still fail to remove temporary outputs of some failed bundles, but at least the common case (where all bundles succeed) is guaranteed to be fully addressed.